File: lib/models/user_model.dart
// user_model.dart
class UserModel {
  final String id;
  final String name;
  final String email;
  final String? mobile;
  final String? username;
  final String? photo;
  final DateTime createdAt;
  final DateTime? passwordChangedAt;
  final bool? active;
  final DateTime? deletedAt;
  final UserStats? stats;
  final List<String>? ownedQuizzes;
  final List<String>? playedQuiz;
  final UserSettings? settings;

  UserModel({
    required this.id,
    required this.name,
    required this.email,
    this.mobile,
    this.username,
    this.photo,
    required this.createdAt,
    this.passwordChangedAt,
    this.active,
    this.deletedAt,
    this.stats,
    this.ownedQuizzes,
    this.playedQuiz,
    this.settings,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) => UserModel(
    id: json['_id'] ?? json['id'],
    name: json['name'],
    email: json['email'],
    mobile: json['mobile']?.toString(),
    username: json['username'],
    photo: json['photo'],
    createdAt: DateTime.parse(json['createdAt']),
    passwordChangedAt: json['passwordChangedAt'] != null
        ? DateTime.parse(json['passwordChangedAt'])
        : null,
    active: json['active'],
    deletedAt: json['deletedAt'] != null
        ? DateTime.parse(json['deletedAt'])
        : null,
    stats: json['stats'] != null ? UserStats.fromJson(json['stats']) : null,
    ownedQuizzes: (json['ownedQuizzes'] as List<dynamic>?)
        ?.map((e) => e.toString()).toList(),
    playedQuiz: (json['playedQuiz'] as List<dynamic>?)
        ?.map((e) => e.toString()).toList(),
    settings: json['settings'] != null
        ? UserSettings.fromJson(json['settings'])
        : null,
  );
}

class UserStats {
  final int totalScore;
  final double averageScore;
  final int gamesPlayed;
  UserStats({
    required this.totalScore,
    required this.averageScore,
    required this.gamesPlayed,
  });
  factory UserStats.fromJson(Map<String, dynamic> json) => UserStats(
    totalScore: json['totalScore'] ?? 0,
    averageScore: (json['averageScore'] ?? 0).toDouble(),
    gamesPlayed: json['gamesPlayed'] ?? 0,
  );
}

class UserSettings {
  final bool darkMode;
  UserSettings({required this.darkMode});
  factory UserSettings.fromJson(Map<String, dynamic> json) =>
      UserSettings(darkMode: json['darkMode'] ?? false);
}

File: Server Models/userModel.js
const crypto = require("crypto");
const mongoose = require("mongoose");
const validator = require("validator");
const bcrypt = require("bcryptjs");

//modeling the database collection
const userSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      require: [true, "User must enter name"],
    },
    email: {
      type: String,
      require: [true, "User must enter email"],
      unique: [true, "Email already in use"],
      lowercase: true,
      validate: [validator.isEmail, "Please enter a valid email"],
    },
    mobile: {
      type: Number,
      validate: {
        validator: function (val) {
          return val && val.length === 10 && /^\d+$/.test(val);
        },
      },
    },
    username: String,
    photo: {
      type: String,
      default: "users/default.jpg",
    },
    password: {
      type: String,
      require: [true, "User must enter password"],
      minlength: 8,
      maxlength: 64,
      select: false,
    },
    passwordConfirm: {
      type: String,
      require: [true, "User must enter password"],
      minlength: 8,
      maxlength: 64,
      validate: {
        validator: function (val) {
          return this.password === val;
        },
        message: "Password and Password Confirm must be same",
      },
    },
    passwordChangedAt: {
      type: Date,
    },
    passwordResetToken: String,
    passwordResetExpires: Date,
    stats: {
      totalScore: {
        type: Number,
        default: 0,
      },
      averageScore: {
        type: Number,
        default: 0,
      },
      gamesPlayed: {
        type: Number,
        default: 0,
      },
    },
    ownedQuizzes: [
      {
        type: mongoose.Schema.ObjectId,
        ref: "Quiz",
      },
    ],
    playedQuiz: [
      {
        type: mongoose.Schema.ObjectId,
        ref: "GameSession",
      },
    ],
    active: {
      type: Boolean,
      default: true,
      select: false,
    },
    deletedAt: {
      type: Date,
    },
    createdAt: {
      type: Date,
      default: Date.now(),
    },
    settings: {
      darkMode: {
        type: Boolean,
        default: false,
      },
      profileVisibility: {
        type: String,
        enum: ["private", "public"],
        required: true,
      },
    },
  },
  {
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  },
);

userSchema.index({ "stats.totalScore": -1 });
userSchema.index({ createdAt: -1 });

//LABEL
// HOOKS
//=> Password Encryption logic
userSchema.pre("save", async function (next) {
  //check if 'password' field is modified or not
  if (!this.isModified("password")) return next();

  //encrypting the password with hash and strength of 12 (higher the strength more CPU intensive and secure the password)
  this.password = await bcrypt.hash(this.password, 12);

  //deleting the passwordConfirm field as it is not necessary for the database (was for user verification only)
  this.passwordConfirm = undefined;
});

//=> set password changed at property dynamically
userSchema.pre("save", function (next) {
  if (!this.isModified("password") || this.isNew) return next();

  this.passwordChangedAt = Date.now() - 1000;
  next();
});

//=> filter to get only active users
userSchema.pre(/^find/g, function (next) {
  this.find({ active: { $ne: false } });
  next();
});

//LABEL
// Methods
//creating an INSTANCE method for all the results objects
//use SchemaName.methods.<method-name>

//=> check if user entered password is correct or not (return false if not same)
userSchema.methods.checkPassword = async function (
  candidatePassword,
  userPassword,
) {
  //candidatePassword = password that is entered by user in forms (from req)
  //userPassword = password that is stored in db (actual password)

  //Use the 'bcrypt' package's compare method, which compares both strings even if either is encrypted
  return await bcrypt.compare(candidatePassword, userPassword);
};

//=> method to check if password is changed after the token generated or not
userSchema.methods.isPasswordChangedAfter = function (JWTTokenTime) {
  if (this.passwordChangedAt) {
    const passwordChangedTime = parseInt(
      this.passwordChangedAt.getTime() / 1000,
    );

    return JWTTokenTime < passwordChangedTime;
  }

  return false;
};

//=> getting the password reset token [simply generating and returning the token]
userSchema.methods.getPasswordResetToken = function () {
  const resetToken = crypto.randomBytes(32).toString("hex");

  this.passwordResetToken = crypto
    .createHash("sha256")
    .update(resetToken)
    .digest("hex");

  this.passwordResetExpires = Date.now() + 10 * 60 * 1000;

  return resetToken;
};

//=> creating the collection
const User = mongoose.model("User", userSchema);

module.exports = User;
